src/app/qgshandlebadlayers.cpp

```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
 {
   return mLayerList->rowCount();
 }
+
+QString QgsHandleBadLayers::findFile( const QString &fileName, const QString &basePath, int maxDepth, int driveMargin )
+{
+  int depth = 0;
+  QDir folder = QDir( QDir( basePath ).absolutePath() );
+  QString existingBase;
+  // find the nearest existing folder
+  while ( !folder.exists() && folder.absolutePath().count( '/' ) > driveMargin )
+  {
+    existingBase = folder.path();
+    if ( !folder.cdUp() )
+      folder = QFileInfo( existingBase ).absoluteDir(); // using fileinfo to move up one level
+    depth += 1;
+  }
+  if ( depth > maxDepth )
+    maxDepth = depth;
+  while ( depth <= maxDepth && folder.cdUp() && folder.absolutePath().count( '/' ) > driveMargin )
```

Comment:
I think the cdUp should be deferred until this loop has run at least once -- otherwise if the move up here results in the driveMargin check failing on the first iteration, you skip checking the actual first found matching folder.

Reply:
Good point, the idea of the limit was to prevent the searching function to go through most of the server/system and bog things down, the limit of 3 level was something that I found convenient for us.

Reply:
I have changed the logic, let me know what you think.